domino/datasets.py

```diff
@@ -0,0 +1,190 @@
+        q = self._create_chunk_queue()
+        with ThreadPoolExecutor(self.max_workers) as executor:
+            # list ensures all the threads are complete before returning results
+            results = list(executor.map(self._upload_chunk, q))
+        return results
+
+    def cancel_upload_session(self):
+        url = self.routes.datasets_cancel_upload(self.dataset_id, self.upload_key)
+        return self.request_manager.get(url)
+
+    def end_upload_session(self):
+        if not self.upload_key:
+            raise RuntimeError(f"upload key for {self.dataset_id} not found. Could not end session.")
+        url = self.routes.datasets_end_upload(self.dataset_id, self.upload_key, self.target_relative_path)
+        return self.request_manager.get(url)
+
+    def _create_chunk_queue(self) -> [UploadChunk]:
+        if os.path.isfile(self.local_path_file_or_directory):
+            return self._create_chunk(self.local_path_file_or_directory)
+        else:
+            nested_q = []
+            for dirpath, dirnames, filenames in os.walk(self.local_path_file_or_directory):
+                for filename in filenames:
+                    # construct the relative path for each file
+                    relative_path_to_file = os.path.join(dirpath, filename)
+                    # append chunk to queue
+                    nested_q.append(self._create_chunk(relative_path_to_file))
+            # flattening list of chunks
+            return [chunk for chunks in nested_q for chunk in chunks]
+
+
+    def _create_chunk(self, local_path_file, starting_index=1):
+        file_size = os.path.getsize(local_path_file)
+        file_name = os.path.basename(local_path_file)
+        total_chunks = max(int(math.ceil(float(file_size) / self.target_chunk_size)), 1)
+        return [UploadChunk(absolute_path=os.path.abspath(local_path_file), chunk_number=chunk_num,
+                            dataset_id=self.dataset_id, file_name=file_name, file_size=file_size,
+                            identifier=f"{file_size}-{file_name}", relative_path=local_path_file,
+                            target_chunk_size=self.target_chunk_size, total_chunks=total_chunks,
+                            upload_key=self.upload_key)
+                for chunk_num in range(starting_index, total_chunks + 1)]
+
+
+    def _upload_chunk(self, chunk: UploadChunk):
+        # read the file chunk
+        starting_skip = chunk.target_chunk_size * (chunk.chunk_number - 1)
+        with open(chunk.absolute_path, 'rb') as file:
+            file.seek(starting_skip)
+            chunk_data = file.read(chunk.target_chunk_size)
+
+        # testing chunk
+        should_upload, checksum = self._test_chunk(chunk, chunk_data)
+
+        # uploading chunk
+        if should_upload:
+            upload_try = 1
+            uploaded = False
+            while upload_try <= MAX_UPLOAD_ATTEMPTS and not uploaded:
```

Comment:
big fan of making good use a decorator for this kind of logic https://pypi.org/project/retry/. This allow clearer code by making the function do only what it is supposed to do